### PR TITLE
Automate publishing crates to crates.io

### PR DESCRIPTION
**Pull request by musicinmybrain**
This mostly reverts 5085c5b038c46ecb9fb16329caf5479b35a963c6, and attempts to support publishing workspace crates `onefetch-ascii`, `onefetch-image`, and `onefetch-manifest`. This would fix https://github.com/o2sh/onefetch/issues/1363 by ensuring crates are published on a platform that fully supports symbolic links.

I have little to no experience with GitHub Workflows, and I am unable to test this PR, but it still *seems* like it should do the job.

### CODE REVIEW

File: .github/workflows/cd.yml [hunk 1]
```diff
@@ -59,5 +59,46 @@ jobs:
         with:
           identifier: o2sh.onefetch
           installers-regex: '\.exe$'
-          version: ${{ github.event.release.tag_name }}
```
Comment (spenserblack) [inline .github/workflows/cd.yml:62]: Why was this line removed?

**Review by spenserblack — CHANGES_REQUESTED** [review]
This looks mostly good. Could you use `cargo publish -p PACKAGE-NAME` instead of setting the working directory? I know I suggested changing the working directories in https://github.com/o2sh/onefetch/commit/5085c5b038c46ecb9fb16329caf5479b35a963c6, but I think using `cargo publish -p PACKAGE-NAME` is slightly cleaner. As an added bonus one could just copy+paste the steps and run them from the project's root to publish manually if we avoid modifying the working directory.

**Comment by spenserblack** [comment]
> I am unable to test this PR

Honestly when it comes to situations like this I think it's fine to basically test in production. In other words, just trigger the workflow when we're ready and expect it to work. If there's anything to fix we can make a new patch release to re-trigger the workflow. I've even seen big projects like Vue make new patches just to re-trigger a previously broken CD.

Though of course we would *never* merge in a broken CD workflow :wink:

File: .github/workflows/cd.yml [hunk 2]
```diff
@@ -59,5 +59,46 @@ jobs:
...
           token: ${{ secrets.WINGET_TOKEN }}
+
+  cargo-publish:
+    name: Publish to crates.io
+    runs-on: ubuntu-latest
+    env:
+      CARGO_TERM_COLOR: always
+
+    steps:
+      - uses: actions/checkout@v3
```
Comment (spenserblack) [inline .github/workflows/cd.yml:71]:
```suggestion
      - uses: actions/checkout@v4
```

File: .github/workflows/cd.yml [hunk 3]
```diff
@@ -59,5 +59,46 @@ jobs:
...
           identifier: o2sh.onefetch
           installers-regex: '\.exe$'
-          version: ${{ github.event.release.tag_name }}
           token: ${{ secrets.WINGET_TOKEN }}
+
+  cargo-publish:
+    name: Publish to crates.io
+    runs-on: ubuntu-latest
+    env:
+      CARGO_TERM_COLOR: always
```
Comment (spenserblack) [inline .github/workflows/cd.yml:68]: I think that this environment variable can actually be set on the workflow level (root level), not the job level. That way it will be set for all jobs, which should help on build steps for other jobs.

File: .github/workflows/cd.yml [hunk 4]
```diff
@@ -59,5 +59,46 @@ jobs:
...
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Install Rust
+        uses: actions-rs/toolchain@v1
+        with:
+          toolchain: stable
+          profile: minimal
+          components: clippy
+
```
Comment (spenserblack) [inline .github/workflows/cd.yml:79]: All of `actions-rs` is deprecated right now. TBH I'm wondering if we actually need a step to install Rust here. GitHub Actions have Rust and `cargo` installed by default, and typically a pretty recent version AFAIK. I never bothered to explicitly install the Rust toolchain in some personal projects, and I've never had a problem.

**Review by spenserblack — CHANGES_REQUESTED** [review]
Sorry, went through this again and had a few more comments.

File: .github/workflows/cd.yml (hunk 1 above)
Comment (musicinmybrain) [reply .github/workflows/cd.yml:62]: It was added in 5085c5b038c46ecb9fb16329caf5479b35a963c6, and then removed again when I reverted that commit. Now, my intention was to revert everything *except the addition of this line*, which is why the commit message says “Revert "fix cd" (but keep the added version)”, but obviously I missed actually fixing up the commit. I’ll correct it. Thanks for catching this.

File: .github/workflows/cd.yml (hunk 3 above)
Comment (musicinmybrain) [reply .github/workflows/cd.yml:68]: I’ll give it a try! This is how it was done in the “old” implementation, but that doesn’t make it optimal. Documentation supports your claim: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#env

File: .github/workflows/cd.yml (hunk 2 above)
Comment (musicinmybrain) [reply .github/workflows/cd.yml:71]: 👍

File: .github/workflows/cd.yml (hunk 4 above)
Comment (musicinmybrain) [reply .github/workflows/cd.yml:79]: Interesting. Looking more closely, the existing `deploy` job has the following “Install Rust” step:

https://github.com/o2sh/onefetch/blob/ac7e2b8e3313c70f7e7960aefa1296e302ac19e6/.github/workflows/cd.yml#L17-L21

I suppose it makes sense to either match that in `cargo-publish`, or remove *both* “Install Rust” steps. What do you think?

**Comment by musicinmybrain** [comment]
> This looks mostly good. Could you use `cargo publish -p PACKAGE-NAME` instead of setting the working directory? I know I suggested changing the working directories in [5085c5b](https://github.com/o2sh/onefetch/commit/5085c5b038c46ecb9fb16329caf5479b35a963c6), but I think using `cargo publish -p PACKAGE-NAME` is slightly cleaner. As an added bonus one could just copy+paste the steps and run them from the project's root to publish manually if we avoid modifying the working directory.

I’ll follow up with this change next. `$ cargo publish -p onefetch-ascii --dry-run` works, so it should be fine.

**Comment by musicinmybrain** [comment]
I think I handled all the feedback except for https://github.com/o2sh/onefetch/pull/1364#discussion_r1674351277. Let me know what you think about https://github.com/o2sh/onefetch/pull/1364#discussion_r1674414547, and I’ll be back to this in a few hours.

File: .github/workflows/cd.yml (hunk 4 above)
Comment (spenserblack) [reply .github/workflows/cd.yml:80]: Let's add dtolnay's action for consistency.

**Comment by spenserblack** [comment]
> I think I handled all the feedback except for [#1364 (comment)](https://github.com/o2sh/onefetch/pull/1364#discussion_r1674351277). Let me know what you think about [#1364 (comment)](https://github.com/o2sh/onefetch/pull/1364#discussion_r1674414547), and I’ll be back to this in a few hours.

Great! Once you resolve https://github.com/o2sh/onefetch/pull/1364#discussion_r1674431906 this will be good to me.

**Comment by musicinmybrain** [comment]
> > I think I handled all the feedback except for [#1364 (comment)](https://github.com/o2sh/onefetch/pull/1364#discussion_r1674351277). Let me know what you think about [#1364 (comment)](https://github.com/o2sh/onefetch/pull/1364#discussion_r1674414547), and I’ll be back to this in a few hours.
> 
> Great! Once you resolve [#1364 (comment)](https://github.com/o2sh/onefetch/pull/1364#discussion_r1674431906) this will be good to me.

Done! Thank you for the thorough and helpful review.

**Review by spenserblack — APPROVED** [review]
Looks good! Congrats on getting some experience with GH Workflows!

**Review by o2sh — APPROVED** [review]
